lib/awesome_print/formatters/array_formatter.rb

```diff
@@ -27,36 +27,104 @@ def format
 
       private
 
-      def methods_array(a)
-        a.sort! { |x, y| x.to_s <=> y.to_s }                  # Can't simply a.sort! because of o.methods << [ :blah ]
-        object = a.instance_variable_get('@__awesome_methods__')
-        tuples = a.map do |name|
-          if name.is_a?(Symbol) || name.is_a?(String)         # Ignore garbage, ex. 42.methods << [ :blah ]
-            tuple = if object.respond_to?(name, true)         # Is this a regular method?
-              the_method = object.method(name) rescue nil     # Avoid potential ArgumentError if object#method is overridden.
-              if the_method && the_method.respond_to?(:arity) # Is this original object#method?
-                method_tuple(the_method)                      # Yes, we are good.
-              end
-            elsif object.respond_to?(:instance_method)              # Is this an unbound method?
-              method_tuple(object.instance_method(name)) rescue nil # Rescue to avoid NameError when the method is not
-            end                                                     # available (ex. File.lchmod on Ubuntu 12).
+      def awesome_array?
+        array.instance_variable_defined?('@__awesome_methods__')
+      end
+
+      def multiline_array?
+        options[:multiline]
+      end
+
+      def empty_array
+        "[]"
+      end
+
+      def simple_array
+        "[ " << array.map{ |item| inspector.awesome(item) }.join(", ") << " ]"
+      end
+
+      def multiline_array
+        width = (array.size - 1).to_s.size
+
+        data = array.inject([]) do |arr, item|
+          index = indent
+          index << colorize("[#{arr.size.to_s.rjust(width)}] ", :array) if options[:index]
+          indented do
+            arr << (index << inspector.awesome(item))
           end
-          tuple || [ name.to_s, '(?)', '?' ]                  # Return WTF default if all the above fails.
         end
 
-        width = (tuples.size - 1).to_s.size
-        name_width = tuples.map { |item| item[0].size }.max || 0
-        args_width = tuples.map { |item| item[1].size }.max || 0
+        data = limited(data, width) if should_be_limited?
+
+        "[\n" << data.join(",\n") << "\n#{outdent}]"
+      end
+
+      def methods_array
+        array.sort! { |x, y| x.to_s <=> y.to_s }                  # Can't simply a.sort! because of o.methods << [ :blah ]
+
+        "[\n#{methods_template}\n#{outdent}]"
+      end
+
+      def tuples
+        @tuples ||= begin
+          object = array.instance_variable_get('@__awesome_methods__')
+
+          tuples = array.map do |name|
```

Comment:
Oh, now I got it. I just move this chunk of code to another method without caring much about it otherwise I would not be able to focus on improve a single method per step.